### PR TITLE
Added --directory to `rpm content -t package upload`.

### LINKED ISSUE #994: Provide easier workflow for uploading rpms into a repository
## Summary
Today with pulp-cli, if a user wants to upload 3 rpms, they have to run:

pulp artifact upload --file "$PKG1"
pulp artifact upload --file "$PKG2"
pulp artifact upload --file "$PKG3"

Then for each of those you have to grab the href of each artifact and fetch the sha256:

pulp show --href "${ARTIFACT_HREF_1}" | jq -r '.sha256')
pulp show --href "${ARTIFACT_HREF_2}" | jq -r '.sha256')
pulp show --href "${ARTIFACT_HREF_3}" | jq -r '.sha256'

Then for each of those, they have to use the sha256 to create a content unit and grab the unit href:

PACKAGE_HREF1=$(pulp rpm content create  --sha256 "${ARTIFACT_SHA256}"  | jq -r '.pulp_href')
PACKAGE_HREF2=$(pulp rpm content create  --sha256 "${ARTIFACT_SHA256}"  | jq -r '.pulp_href')
PACKAGE_HREF3=$(pulp rpm content create  --sha256 "${ARTIFACT_SHA256}"  | jq -r '.pulp_href')

Finally you can add these to a repository:

TASK_HREF=$(pulp rpm repository content modify \
            --repository "${REPO_NAME}" \
            --add-content "[{\"pulp_href\": \"${PACKAGE_HREF1}\"}, {\"pulp_href\": \"${PACKAGE_HREF2}\"}, {\"pulp_href\": \"${PACKAGE_HREF3}\"}]" \
            2>&1 >/dev/null | awk '{print $4}')

This is a LOT of commands to run for a user and there isn't an easy way for me to easily give a couple commands to a user to run to upload an arbitrary list of rpms.  Having a command that handled everything and simply took in a list of rpms and outputted a repo version would be much simpler.

An example of what i'm thinking of:

```
# pulp rpm repository content upload  --repository "${REPO_NAME}"  --files=*.rpm

A new repository version has been created at  /pulp/api/v3/repositories/rpm/abcdef/versions/5
```

This command would need to handle things such as:
* An rpm already exists (such as if i interrupt the command half way through and restart it)

### PR DESCRIPTION
**Pull request by ggainey**
This finds all *.rpm in the specified directory and arranges for them to be added to the specified --repository, and then publishes the final resulting repository-version.

closes #994.

### CODE REVIEW

**Comment by ggainey** [comment]
This is a modification of the approach in https://github.com/pulp/pulp-cli/pull/1000 that does not rely on creating a tmp-repository.

**Comment by ggainey** [comment]
Sample output:
```
$ pulp rpm content upload --directory ./signed/fixtures.pulpproject.org/rpm-signed/ --repository rpm:rpm:upload
[About to upload 35 files for upload.
Started background task /pulp/api/v3/tasks/01905b60-998f-7997-b947-d8145e8800ab/
Done.
Uploaded ./signed/fixtures.pulpproject.org/rpm-signed/crow-0.8-1.noarch.rpm...
...
Created new version 35 in upload
Publishing repository upload.
Started background task /pulp/api/v3/tasks/01905b61-58ff-7909-81ef-84cd74d8b42f/
Done.
{
  "pulp_href": "/pulp/api/v3/publications/rpm/rpm/01905b61-5958-7670-9d88-9b486ec25d8b/",
  "pulp_created": "2024-06-27T20:27:04.921566Z",
  "pulp_last_updated": "2024-06-27T20:27:05.132778Z",
  "repository_version": "/pulp/api/v3/repositories/rpm/rpm/01905b60-7c70-75d6-8882-0ec1a29fa71d/versions/35/",
  "repository": "/pulp/api/v3/repositories/rpm/rpm/01905b60-7c70-75d6-8882-0ec1a29fa71d/",
  "checksum_type": "sha256",
  "metadata_checksum_type": "sha256",
  "package_checksum_type": "sha256",
  "gpgcheck": null,
  "repo_gpgcheck": null,
  "sqlite_metadata": false,
  "repo_config": {},
  "compression_type": null
}
```

**Comment by ggainey** [comment]
List of REST API calls made in response to using this PR to upload a directory containing one RPM:

- Find the named repository                                                                  
  - **repositories_rpm_rpm_list** : GET http://localhost:5001/pulp/api/v3/repositories/rpm/rpm/?name=upload&offset=0&limit=1
- Create a package "into" it                                                                 
  - **content_rpm_packages_create** : POST http://localhost:5001/pulp/api/v3/content/rpm/packages/ 
- Waiting for task-completion                                                                
  - **tasks_read** : GET http://localhost:5001/pulp/api/v3/tasks/01905f70-962f-73c0-b718-5b37d6fb3a3b/
- Verify the resulting content                                                               
  - **content_rpm_packages_read** : GET http://localhost:5001/pulp/api/v3/content/rpm/packages/01904ff9-5f44-7c91-a7be-f93bbd90f77d/
- Finding the named repository's latest-version once we're done uploading                    
  - **repositories_rpm_rpm_read** : GET http://localhost:5001/pulp/api/v3/repositories/rpm/rpm/01905b60-7c70-75d6-8882-0ec1a29fa71d/
- Publishing that version                                                                    
  - **publications_rpm_rpm_create** : POST http://localhost:5001/pulp/api/v3/publications/rpm/rpm/ 
- Spitting out the resulting publication                                                     
  - **publications_rpm_rpm_read** : GET http://localhost:5001/pulp/api/v3/publications/rpm/rpm/01905f70-a352-7905-9d04-bd31c7615b56/

**Comment by mdellweg** [comment]
Uploads api's are only employed when the file is bigger than the specified chunk size.
Please try another run with chunksize 2/3 the file size.

**Comment by ggainey** [comment]
> Uploads api's are only employed when the file is bigger than the specified chunk size.
> Please try another run with chunksize 2/3 the file size.

```
pulp -vvv rpm content upload --directory ./test_upload --repository rpm:rpm:upload --chunk-size 1KB
```

adds the following to the above-listed REST calls:
- create an Upload
  - **uploads_create** : POST http://localhost:5001/pulp/api/v3/uploads/
- read the result back
  - **uploads_read** : GET http://localhost:5001/pulp/api/v3/uploads/019093af-3c0f-7093-ade9-5b520e68a930/
- Update the Upload one chunk at a time
   - **uploads_update** : PUT http://localhost:5001/pulp/api/v3/uploads/019093af-3c0f-7093-ade9-5b520e68a930/

**Comment by ggainey** [comment]
This now combines the behavior of #1000 and #1003 - you can get a single repo-version by specifying `--use-temp-repository`, and ask for a publication to the specified `--repository` with `--publish`. The defaults are chosen to match the original request - upload directly to the specified repository, and publish when done.

Example output:
```
$ pulp rpm repository create --name upload2 --no-autopublish
{
  "pulp_href": "/pulp/api/v3/repositories/rpm/rpm/0190a752-e905-7773-bf51-f9041904fdd8/",
  "pulp_created": "2024-07-12T14:22:27.078395Z",
  "pulp_last_updated": "2024-07-12T14:22:27.084739Z",
  "versions_href": "/pulp/api/v3/repositories/rpm/rpm/0190a752-e905-7773-bf51-f9041904fdd8/versions/",
  "pulp_labels": {},
  "latest_version_href": "/pulp/api/v3/repositories/rpm/rpm/0190a752-e905-7773-bf51-f9041904fdd8/versions/0/",
  "name": "upload2",
  "description": null,
  "retain_repo_versions": null,
  "remote": null,
  "autopublish": false,
  "metadata_signing_service": null,
  "package_signing_service": null,
  "package_signing_fingerprint": "",
  "retain_package_versions": 0,
  "checksum_type": null,
  "metadata_checksum_type": null,
  "package_checksum_type": null,
  "gpgcheck": null,
  "repo_gpgcheck": null,
  "sqlite_metadata": false,
  "repo_config": {},
  "compression_type": null
}
$ pulp rpm content upload --directory ./signed/fixtures.pulpproject.org/rpm-signed/ --repository upload2
About to upload 35 files for upload2.
Started background task /pulp/api/v3/tasks/0190a753-0cdd-739d-b60b-9e952120411c/
Done.
Uploaded ./signed/fixtures.pulpproject.org/rpm-signed/crow-0.8-1.noarch.rpm...
...
Started background task /pulp/api/v3/tasks/0190a753-c2e9-7d6f-bc9e-fd736a608c9d/
Done.
Uploaded ./signed/fixtures.pulpproject.org/rpm-signed/camel-0.1-1.noarch.rpm...
Created new version 35 in upload2
Publishing repository upload2.
Started background task /pulp/api/v3/tasks/0190a753-c865-7684-aea7-fb5b64257b30/
Done.
{
  "pulp_href": "/pulp/api/v3/publications/rpm/rpm/0190a753-c8b6-7bae-9c9f-5f0ebc7d0e5a/",
  "pulp_created": "2024-07-12T14:23:24.343140Z",
  "pulp_last_updated": "2024-07-12T14:23:24.527520Z",
  "repository_version": "/pulp/api/v3/repositories/rpm/rpm/0190a752-e905-7773-bf51-f9041904fdd8/versions/35/",
  "repository": "/pulp/api/v3/repositories/rpm/rpm/0190a752-e905-7773-bf51-f9041904fdd8/",
  "checksum_type": "sha256",
  "metadata_checksum_type": "sha256",
  "package_checksum_type": "sha256",
  "gpgcheck": null,
  "repo_gpgcheck": null,
  "sqlite_metadata": false,
  "repo_config": {},
  "compression_type": null
}
$ pulp rpm publication list --repository upload2
[
  {
    "pulp_href": "/pulp/api/v3/publications/rpm/rpm/0190a753-c8b6-7bae-9c9f-5f0ebc7d0e5a/",
    "pulp_created": "2024-07-12T14:23:24.343140Z",
    "pulp_last_updated": "2024-07-12T14:23:24.527520Z",
    "repository_version": "/pulp/api/v3/repositories/rpm/rpm/0190a752-e905-7773-bf51-f9041904fdd8/versions/35/",
    "repository": "/pulp/api/v3/repositories/rpm/rpm/0190a752-e905-7773-bf51-f9041904fdd8/",
    "checksum_type": "sha256",
    "metadata_checksum_type": "sha256",
    "package_checksum_type": "sha256",
    "gpgcheck": null,
    "repo_gpgcheck": null,
    "sqlite_metadata": false,
    "repo_config": {},
    "compression_type": null
  }
]
```

**Comment by ggainey** [comment]
NOTE: at some point, we may want to modify to allow multiple --directory --file to be specified. In the moment, decided it was better to answer the immediate known-requirement, and leave that as a Future Enhancement.

File: CHANGES/994.feature [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Expanded on options available to `rpm content -t package upload`.
+
+The user can now:
+  * Upload an entire directory of RPMs
+  * Choose to have them added to their desired destionation as a
+    single new repository-version.
+  * Choose to have a new Publication created at the end of the
+    uploads.
```
Comment (mdellweg) [inline CHANGES/994.feature:8]:
```suggestion
The user can now:
  * Upload an entire directory of RPMs
  * Choose to have them added to their desired destination as a single new repository-version.
  * Choose to have a new Publication created at the end of the uploads.
```

Comment (mdellweg) [reply CHANGES/994.feature:8]: Oh why....
Note: apart from the obvious unnecessary linebreaks, there was a typo in `destiOnation`.

File: pulpcore/cli/rpm/content.py [hunk 2]
```diff
@@ -285,13 +323,152 @@ def upload(
...
+                _("You must specify a --repository to use --directory uploads.")
+            )
+
+        # Sanity: ignore publish|use_temp unless directory has been specified
+        use_tmp = final_dest_repo_ctx and kwargs["use_temp_repository"]
+        do_publish = final_dest_repo_ctx and kwargs["publish"]
+
+        # Sanity: ignore relative_path if directory specified
+        if directory and kwargs["relative_path"]:
+            print(_("relative_path ignored on --directory uploads."))
```
Comment (mdellweg) [inline pulpcore/cli/rpm/content.py:348]: Why not fail on this one?
Also i think it should be `click.echo` (don't ask my why...).

File: pulpcore/cli/rpm/content.py [hunk 3]
```diff
@@ -285,13 +323,152 @@ def upload(
...
+                if use_tmp:
+                    result = _copy_to_final(dest_repo_ctx, final_dest_repo_ctx, pulp_ctx)
+            finally:
+                if use_tmp and dest_repo_ctx:
+                    # Remove the tmp-upload repo
+                    dest_repo_ctx.delete()
+
+        final_version_number = _latest_version_number(final_dest_repo_ctx)
+        if final_version_number:
+            print(
```
Comment (mdellweg) [inline pulpcore/cli/rpm/content.py:372]: Again. Better use `click.echo(..., err=True)`, because the output on stdout should match the `--format` option.

**Review by mdellweg — APPROVED** [review]

**Pull request merged**